### PR TITLE
LCAM-1430|Enhance error handling to pass the MAAT Application error

### PR DESCRIPTION
**Pull request by venkat980**
## What

[Link to story](https://dsdmoj.atlassian.net/browse/LCAM-1430)

Enhance error handling to pass the MAAT Application error so that the exception detail is not masked and can be displayed on the UI.

### CODE REVIEW

File: crime-commons-spring-boot-starter-rest-client/src/main/java/uk/gov/justice/laa/crime/commons/filters/WebClientFilters.java [hunk 1]
```diff
@@ -86,29 +88,33 @@ public static ExchangeFilterFunction handleErrorResponse() {
...
+                        }
+                        return Mono.error(new HttpClientErrorException(response.statusCode(), errorMessage));
                     }
 
                     if (status.is5xxServerError()) {
-                        return new HttpServerErrorException(r.statusCode(), errorMessage);
+                        Optional<Mono<ErrorDTO>> errorDTOMono = Optional.ofNullable(response.bodyToMono(ErrorDTO.class));
+                        if (HttpStatus.INTERNAL_SERVER_ERROR == response.statusCode() && errorDTOMono.isPresent()) {
+                            return errorDTOMono.get()
+                                    .flatMap(errorBody -> Mono.error(new MAATApplicationException(errorBody.getMessage())));
```
Comment (nigelpain) [inline crime-commons-spring-boot-starter-rest-client/src/main/java/uk/gov/justice/laa/crime/commons/filters/WebClientFilters.java:113]: Wont this filter apply to calls made between microservices aswell, for example c3 -> hardship in which case is MaatApplicationException the error we should be creating here or should it be more generic such as HTTPClientErrorException?

File: crime-commons-spring-boot-starter-rest-client/src/main/java/uk/gov/justice/laa/crime/commons/common/ErrorDTO.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+@Value
+@EqualsAndHashCode(callSuper = false)
+@AllArgsConstructor
+@NoArgsConstructor(force = true)
+@Builder
+public class ErrorDTO extends Throwable {
+    String traceId;
+    String code;
+    String message;
+}
```
Comment (nigelpain) [inline crime-commons-spring-boot-starter-rest-client/src/main/java/uk/gov/justice/laa/crime/commons/common/ErrorDTO.java:18]: Whats the reason for not using the ErrorDTO in commons classes? Is it to try and avoid dependencies between the sub modules of commons library?

Comment (venkat980) [reply crime-commons-spring-boot-starter-rest-client/src/main/java/uk/gov/justice/laa/crime/commons/common/ErrorDTO.java:18]: Correct

File: crime-commons-spring-boot-starter-rest-client/src/main/java/uk/gov/justice/laa/crime/commons/filters/WebClientFilters.java (hunk 1 above)
Comment (venkat980) [reply crime-commons-spring-boot-starter-rest-client/src/main/java/uk/gov/justice/laa/crime/commons/filters/WebClientFilters.java:113]: As of now we are only dealing with the MAATApplicationException returned from the Court Data API, to limit the changes.

File: crime-commons-spring-boot-starter-rest-client/src/main/java/uk/gov/justice/laa/crime/commons/common/ErrorDTO.java (hunk 2 above)
Comment (nigelpain) [reply crime-commons-spring-boot-starter-rest-client/src/main/java/uk/gov/justice/laa/crime/commons/common/ErrorDTO.java:18]: Ok thats fine, just wanted to double check 😄

File: crime-commons-spring-boot-starter-rest-client/src/main/java/uk/gov/justice/laa/crime/commons/filters/WebClientFilters.java (hunk 1 above)
Comment (nigelpain) [reply crime-commons-spring-boot-starter-rest-client/src/main/java/uk/gov/justice/laa/crime/commons/filters/WebClientFilters.java:113]: Oh ok so we can expand on this later, that makes sense thanks 👍

**Review by nigelpain — DISMISSED** [review]

**Review by Mhowell494 — APPROVED** [review]

**Pull request merged**